### PR TITLE
bpo-34790: Implement deprecation of passing coroutines to asyncio.wait()

### PR DESCRIPTION
**Pull request by aeros**
/cc @1st1 @asvetlov 

https://bugs.python.org/issue34790

### CODE REVIEW

**Comment by aeros** [comment]
Adding a `DO-NOT-MERGE` label for now. Although the deprecation seems to be properly implemented at the moment, I have to refactor the existing tests with several ` self.assertWarns(DeprecationWarning)` where coroutines are being passed to `asyncio.wait()`.

**Comment by aeros** [comment]
>  I have to refactor the existing tests with self.assertWarns(DeprecationWarning) where coroutines are being passed to asyncio.wait()

Done, removing `DO-NOT-MERGE` label. 

[aeros:~/repos/aeros-cpython]$ ./python -W error -m test test_asyncio
0:00:00 load avg: 0.33 Run tests sequentially
0:00:00 load avg: 0.33 [1/1] test_asyncio
test_asyncio passed in 43.2 sec

== Tests result: SUCCESS ==

1 test OK.

Total duration: 43.2 sec
Tests result: SUCCESS

File: Lib/test/test_asyncio/test_tasks.py
Comment (aeros) [inline Lib/test/test_asyncio/test_tasks.py:1351]: The comment seemed necessary here because it would likely not be initially clear why this should cause a deprecation warning, particularly in the future when the tests have to be modified in 3.11. At a first glance, one may assume that `futs` is a list of Futures based on it's name alone. However, `futs` is actually a list of completed coroutine objects:

```python
>>> import asyncio
>>> import asyncio.coroutines as coroutines
>>> a = asyncio.sleep(0)
>>> b = asyncio.sleep(0)
>>> fs = {a, b}
>>> futs = list(asyncio.as_completed(fs))
>>> for f in futs:
...   coroutines.iscoroutine(f)
... 
True
True
>>> for f in futs:
...    f
<coroutine object as_completed.<locals>._wait_for_one at 0x7f056fb4bdc0>
<coroutine object as_completed.<locals>._wait_for_one at 0x7f056fb80dc0>
```
This makes more sense when the exact behavior of `asyncio.as_completed()` is taken into consideration, but I think it would be reasonable to change the variable name `futs` into `completed_coros`. IMO, this would make it a lot less confusing at virtually no expense.

Admittedly, this took a fair bit of troubleshooting on my end before I realized what was going on. `asyncio.completed_coros()` is not exactly a function I use on a day-to-day basis, so I had not initially realized that `futs` was a list of completed coroutine objects.

What made this a bit more difficult to figure out was that the [documentation](https://docs.python.org/3/library/asyncio-task.html#asyncio.as_completed) for `asyncio.as_completed()` makes no mention of coroutine objects being returned as completed coroutine objects. The only mention of this behavior is in the [docstring](https://github.com/python/cpython/blob/0078a0c2a59a358fa032ec9847c108378b66b656/Lib/asyncio/tasks.py#L556):
```
    When waiting for the yielded coroutines you'll get the results (or
    exceptions!) of the original Futures (or coroutines), in the order
    in which and as soon as they complete.
```
I don't know that this situation is an especially regular occurrence, but I think it would be worth clarifying this in the documentation.

**Review by asvetlov — APPROVED** [review]
LGTM.
The PR **should not** be backported to 3.8, right?
It raises new deprecations that were absent in 3.8.0; adding them in 3.8.1 is confusing.

**Comment by aeros** [comment]
@asvetlov 
> The PR should not be backported to 3.8, right?

Yep, that's why I did this one separately from https://github.com/python/cpython/pull/16975, which only fixed the 3.8 whatsnew entry based on Yury's suggestion. Otherwise, the changes would've been combined into the same PR.

>  It raises new deprecations that were absent in 3.8.0; adding them in 3.8.1 is confusing.

Agreed, I'm starting to realize that we should try to avoid implementing significant deprecations partway through a major version (3.x) when possible. It would more likely comely across to users as a surprise and lead to some confusion rather than being overly helpful. For example, if users have already ran their tests with warnings enabled and resolved any deprecation warnings for 3.8.0, I don't think it would be a great experience for them to encounter a bunch of new ones in 3.8.1.

**Comment by aeros** [comment]
@1st1 @asvetlov What are your thoughts on updating the [documentation](https://docs.python.org/3/library/asyncio-task.html#asyncio.as_completed) of `asyncio.as_completed()` to mention that an iterator of completed coroutine objects are returned when coroutine objects are passed to *fs*? I don't think this behavior is very obvious. 

For more context, see https://github.com/python/cpython/pull/16977#discussion_r339919277.

**Comment by aeros** [comment]
@asvetlov @1st1 Is there anything else that needs to added to this PR or are we just waiting for a final review?

**Comment by asvetlov** [comment]
Thanks.
Regarding mentioned docs for `as_completed()` and `ensure_future()` -- please let's discuss in separate issues.
I quite don't follow your proposals, sorry.

**Comment by aeros** [comment]
>  Regarding mentioned docs for as_completed() and ensure_future() -- please let's discuss in separate issues.

Yeah, I'll have to spend more time fleshing those ideas out and consider if the proposals might be worthwhile.

Thanks!

**Pull request merged**
